src/java.base/share/classes/java/lang/template/StringTemplate.java

```diff
@@ -0,0 +1,351 @@
+                String.join("\", \"", stringTemplate.fragments()) +
+                "\" ], values = " +
+                stringTemplate.values() +
+                " }";
+    }
+
+    /**
+     * Produces a hash code for the supplied {@link StringTemplate}.
+     *
+     * @param stringTemplate the {@link StringTemplate} to hash
+     *
+     * @return hash code for the supplied {@link StringTemplate}
+     *
+     * @throws NullPointerException if stringTemplate is null
+     *
+     * @implSpec The hashCode is the bit XOR of the fragments hash code and
+     * the values hash code.
+     */
+    public static int hashCode(StringTemplate stringTemplate) {
+        Objects.requireNonNull(stringTemplate, "stringTemplate should not be null");
+        return Objects.hash(stringTemplate.fragments(), stringTemplate.values());
+    }
+
+    /**
+     * Tests equality of two {@link StringTemplate}.
+     *
+     * @param a  first {@link StringTemplate}
+     * @param b  second {@link StringTemplate}
+     *
+     * @return true if the two {@link StringTemplate StringTemplates} are equivalent
+     *
+     * @throws NullPointerException if either @link StringTemplate} is null
+     *
+     * @implSpec Equality is determined by testing equality of the fragments and
+     * the values.
+     */
+    public static boolean equals(Object a, Object b) {
+        Objects.requireNonNull(a, "StringTemplate a should not be null");
+        Objects.requireNonNull(b, "StringTemplate b should not be null");
+        return a instanceof StringTemplate aST && b instanceof StringTemplate bST &&
+                Objects.equals(aST.fragments(), bST.fragments()) &&
+                Objects.equals(aST.values(), bST.values());
+    }
+
+    /**
+     * Returns a StringTemplate composed from a string.
+     *
+     * @param string  single string fragment
+     *
+     * @return StringTemplate composed from string
+     *
+     * @throws NullPointerException if string is null
+     */
+    public static StringTemplate of(String string) {
+        Objects.requireNonNull(string, "string must not be null");
+        return new TemplateRuntime.SimpleStringTemplate(List.of(string), List.of());
+    }
+
+    /**
+     * Returns a StringTemplate composed from fragments and values.
+     *
+     * @implSpec The {@code fragments} list size must be one more that the
+     * {@code values} list size.
+     *
+     * @param fragments list of string fragments
+     * @param values    list of expression values
+     *
+     * @return StringTemplate composed from string
+     *
+     * @throws IllegalArgumentException if fragments list size is not one more
+     *         than values list size
+     * @throws NullPointerException if fragments is null or values is null or if any fragment is null.
+     *
+     * @implNote Contents of both lists are copied to construct immutable lists.
+     */
+    public static StringTemplate of(List<String> fragments, List<Object> values) {
```

Comment:
Should be `StringTemplate of(List<String> fragments, List<?> values) {`

The call to List.copyOf() will change the List<?> to List<Object>.

Reply:
> raw List ?

oops, formatting issue.

The idea is that `values` can be typed `List<?>` because the call to List.copyOf() will take the `List<?>` and return a `List<Object>`.

And as a type of a parameter `List<?>` is better than `List<Object>` because `List<String>` is a subtype of `List<?>` but not a subtype of `List<Object>`.

Reply:
Got you. Changing.